janusgraph-core/src/main/java/org/janusgraph/graphdb/util/SubqueryIterator.java

```diff
@@ -88,9 +86,13 @@ public JanusGraphElement next() {
     }
 
     @Override
-    public void close() throws Exception {
+    public void close() {
```

Comment:
I didn't spend much time checking the logic behind this class but from the first glance both solutions look error prone to me.
The first solution binds the logic to `hasNext` to add new elements to the cache (no one said that `hasNext` should always be executed before `next` or `close` methods, thus the solution is error prone).
A new solution binds logic to `close` method which might be a little bit better but still looks error prone to me (if we don't call `close` method at all). 
I think it would be more correct to bind that logic to `next` method because only `next` method is responsible for retrieving new elements. Without calling `next` we can't collect necessary elements for our cache.
I didn't had much time to think about another solution but what if we use a separate method, something like:
```
private void cacheQueryIfNoMoreElementsAvailable(){
    if (!elementIterator.hasNext() && currentIds != null) {
        indexCache.put(subQuery, currentIds);
        profiler.setResultSize(currentIds.size());
        profiler.stopTimer();
        isTimerRunning = false;
    }
}
```
Than our next method could look like:
```
public JanusGraphElement next() {
    JanusGraphElement element = this.elementIterator.next();
    cacheQueryIfNoMoreElementsAvailable();
    return element;
}
```
In addition we can execute this method in the end of `SubqueryIterator` constructor just for cases when an index contains 0 elements.
That said, I didn't check the above soltuion and didn't think too much about it, thus I may missed something. Nevertheless, I would like to hear some opinions regarding it.

Reply:
Nice idea. I think stopping the timer in the close() method still would be useful in the try-with-resources statement, or when the user quits the iteration (although I cannot think of such a case where profiling is used). I think I might let next() method call hasNext() method, which in turn calls close() method when results are exhausted.

It also comes into my mind that the current solution has a bug: when the iteration is closed before iterating is over, we shouldn't put results into the cache, since the results are probably incomplete.

Reply:
Agree, so I assume the resulting code could look something like:
```
public class SubqueryIterator implements CloseableIterator<JanusGraphElement> {

    private final JointIndexQuery.Subquery subQuery;

    private final Cache<JointIndexQuery.Subquery, List<Object>> indexCache;

    private final Iterator<? extends JanusGraphElement> elementIterator;

    private List<Object> currentIds;

    private QueryProfiler profiler;

    private boolean isTimerRunning;

    public SubqueryIterator(JointIndexQuery.Subquery subQuery, IndexSerializer indexSerializer, BackendTransaction tx,
            Cache<JointIndexQuery.Subquery, List<Object>> indexCache, int limit,
            Function<Object, ? extends JanusGraphElement> function, List<Object> otherResults) {
        this.subQuery = subQuery;
        this.indexCache = indexCache;
        final List<Object> cacheResponse = indexCache.getIfPresent(subQuery);
        final Stream<?> stream;
        if (cacheResponse != null) {
            stream = cacheResponse.stream();
        } else {
            try {
                currentIds = new ArrayList<>();
                profiler = QueryProfiler.startProfile(subQuery.getProfiler(), subQuery);
                isTimerRunning = true;
                stream = indexSerializer.query(subQuery, tx).peek(r -> currentIds.add(r));
            } catch (final Exception e) {
                throw new JanusGraphException("Could not call index", e.getCause());
            }
        }
        elementIterator = stream.filter(e -> otherResults == null || otherResults.contains(e)).limit(limit).map(function).map(r -> (JanusGraphElement) r).iterator();
        cacheQueryIfNoMoreElementsAvailable();
    }

    @Override
    public boolean hasNext() {
        return elementIterator.hasNext();
    }

    @Override
    public JanusGraphElement next() {
        JanusGraphElement element = this.elementIterator.next();
        cacheQueryIfNoMoreElementsAvailable();
        return element;
    }

    @Override
    public void close() {
        if (isTimerRunning) {
            profiler.stopTimer();
        }
    }

    private void cacheQueryIfNoMoreElementsAvailable(){
        if (!elementIterator.hasNext() && currentIds != null) {
            indexCache.put(subQuery, currentIds);
            profiler.setResultSize(currentIds.size());
            profiler.stopTimer();
            isTimerRunning = false;
        }
    }
}
```